nexus_system_orchestrator/src/context.hpp

```diff
@@ -29,47 +29,234 @@
+    return *this;
+  }
+
+  std::string get_task_results() const
+  {
+    std::lock_guard<std::mutex> lock(_task_results_mutex);
+    return _task_results;
+  }
+
+  Context& add_queued_signal(const std::string& task_id, const std::string& signal) {
+    std::lock_guard<std::mutex> lock(_queued_signals_mutex);
+    _queued_signals[task_id].emplace_back(signal);
+    return *this;
+  }
+
+  std::unordered_map<std::string, std::vector<std::string>> get_queued_signals() const
+  {
+    std::lock_guard<std::mutex> lock(_queued_signals_mutex);
+    return _queued_signals;
+  }
+
+private:
   // using reference to prevent circular references
-public: rclcpp_lifecycle::LifecycleNode& node;
-public: std::string job_id;
-public: WorkOrder wo;
-public: std::vector<WorkcellTask> tasks;
-public: std::shared_ptr<const common::TaskRemapper> task_remapper;
+  rclcpp_lifecycle::LifecycleNode& _node;
+  std::string _job_id;
+  WorkOrder _wo;
+  std::vector<WorkcellTask> _tasks;
+  std::shared_ptr<const common::TaskRemapper> _task_remapper;
+
   /**
    * Map of task ids and their assigned workcell ids.
    */
-public: std::unordered_map<std::string, std::string> workcell_task_assignments;
+  std::unordered_map<std::string, std::string> _workcell_task_assignments = {};
 
   /**
    * Map of workcell ids and it's session.
    */
-public: std::unordered_map<std::string,
-    std::shared_ptr<WorkcellSession>> workcell_sessions;
+  std::unordered_map<std::string,
+  std::shared_ptr<WorkcellSession>> _workcell_sessions;
   /**
    * Map of transporter ids and it's session.
    */
-public: std::unordered_map<std::string,
-    std::shared_ptr<TransporterSession>> transporter_sessions;
-public: std::unordered_map<std::string, TaskState> task_states;
-public: std::shared_ptr<WorkOrderGoalHandle> goal_handle;
-public: std::vector<std::string> errors;
-public: std::string task_results;
+  std::unordered_map<std::string,
+  std::shared_ptr<TransporterSession>> _transporter_sessions;
+  std::unordered_map<std::string, TaskState> _task_states = {};
+  std::shared_ptr<WorkOrderGoalHandle> _goal_handle = nullptr;
+  std::vector<std::string> _errors = {};
+  std::string _task_results;
 
   /**
    * Map of workcell task ids and their queued signals.
    */
-public: std::unordered_map<std::string,
-    std::vector<std::string>> queued_signals;
+  std::unordered_map<std::string, std::vector<std::string>> _queued_signals;
+
+  mutable std::mutex _node_mutex;
+  mutable std::mutex _job_id_mutex;
+  mutable std::mutex _wo_mutex;
+  mutable std::mutex _tasks_mutex;
+  mutable std::mutex _task_remapper_mutex;
+  mutable std::mutex _workcell_task_assignments_mutex;
+  mutable std::mutex _workcell_sessions_mutex;
+  mutable std::mutex _transporter_sessions_mutex;
+  mutable std::mutex _task_states_mutex;
+  mutable std::mutex _goal_handle_mutex;
+  mutable std::mutex _errors_mutex;
+  mutable std::mutex _task_results_mutex;
+  mutable std::mutex _queued_signals_mutex;
```

Comment:
Why not lock a single mutex to access all members?

Reply:
https://github.com/osrf/nexus/pull/75/commits/c9448b1458351abf08290421c628072ff4c6d93b